logstash-core/src/main/java/org/logstash/common/clients/HttpClient.java

```diff
@@ -0,0 +1,655 @@
+        request.setEntity(new StringEntity(requestBody));
+
+        if (headers != null) {
+            headers.forEach(request::addHeader);
+        }
+
+        try (CloseableHttpResponse response = httpClient.execute(request)) {
+            if (response.getStatusLine().getStatusCode() >= 400) {
+                throw new RequestFailedException("HEAD", url , response.getStatusLine().getReasonPhrase());
+            }
+
+            ByteArrayOutputStream responseBody = new ByteArrayOutputStream();
+            response
+                    .getEntity()
+                    .writeTo(responseBody);
+            return responseBody.toString("UTF-8");
+        } catch (IOException e) {
+            throw new RequestFailedException("PUT", url, e);
+        }
+    }
+
+    /**
+     * Performs an HTTP DELETE request
+     *
+     * @param relativePath  Relative path to resource, e.g. api/status
+     * @throws RequestFailedException
+     * @throws IOException
+     */
+    public void delete(String relativePath) throws RequestFailedException, IOException {
+        delete(relativePath, null);
+    }
+
+    /**
+     * Performs an HTTP DELETE request
+     *
+     * @param relativePath  Relative path to resource, e.g. api/status
+     * @param headers       Headers to include with request
+     * @throws RequestFailedException
+     * @throws IOException
+     */
+    public void delete(String relativePath, Map<String, String> headers) throws RequestFailedException, IOException {
+        String url = makeUrlFrom(relativePath);
+        HttpDelete request = new HttpDelete(url);
+
+        if (headers != null) {
+            headers.forEach(request::addHeader);
+        }
+
+        try(CloseableHttpResponse response = httpClient.execute(request)) {
+            if (response.getStatusLine().getStatusCode() >= 400) {
+                throw new RequestFailedException("HEAD", url , response.getStatusLine().getReasonPhrase());
+            }
+        } catch (IOException e) {
+            throw new RequestFailedException("DELETE", url ,e);
+        }
+    }
+
+    private String makeUrlFrom(String relativePath) {
+        return this.baseUrl.toString().replaceFirst("/$", "")
+                + '/'
+                + relativePath.replaceFirst("^/", "");
+    }
+
+    /**
+     * Build an instance of the HTTP client with default configuration options.
+     *
+     * @return HTTP client instance
+     * @throws OptionsBuilderException
+     */
+    public static HttpClient build() throws OptionsBuilderException {
+        return new OptionsBuilder().build();
+    }
+
+    /**
+     * Start building an instance of the HTTP client.
+     *
+     * @return An OptionsBuilder instance to allow configuration of the HTTP client before building it.
+     */
+    public static OptionsBuilder builder() {
+        return new OptionsBuilder();
+    }
+
+    /**
+     * Builder pattern obj
+     */
+    public static class OptionsBuilder {
```

Comment:
I love this options builder, and I love how it encapsulates things I've done for years with a much cleaner API